File: Exams/Resoluções de Exames by Carlos Martins/synchs-async/CountDownLatchAsync.java
/**
 *
 *  ISEL, LEIC, Concurrent Programming
 *
 *  CountDownLatch with asynchronous and synchronous interfaces
 *
 *  Carlos Martins, December 2019
 *
 **/

import java.util.LinkedList;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;

public class CountDownLatchAsync {
			
	// Type used to represent each asynchronous waiter
	private class AsyncWaiter extends CompletableFuture<Boolean> implements Runnable {
		private ScheduledFuture<?> timer;	// the timeout timer, if any
		private final AtomicBoolean lock = new AtomicBoolean(false);	// the lock

		/**
		 * This is the timeout cancellation handler
		 */
		@Override
		public void run() {
			if (tryLock()) {
				// We must acquire the lock in order to try to remove the
				// request from the queue
				synchronized(theLock) {
					asyncWaiters.remove(this);
				}
				// Release resources and complete CF<> with false result.
				close();
				complete(false);
			}
		}

		/**
		 * Tries to lock the waiter in order to satisfy or cancel it.
		 */
		boolean tryLock() {
			return !lock.get() && lock.compareAndSet(false, true);
		}
		/**
		 * Disposes the resources associated with the async acquire
		 */
		void close() {
			if (timer != null)
				timer.cancel(false);
		}
	}

	// The lock - we do not use the monitor functionality
	private final Object theLock = new Object();
	
	// The initial and current count
	private final int initialCount;
	private final AtomicInteger count;

	// The state of the latch	
	private volatile boolean open;		// volatile grants visibility w/o acquire/release the lock

	// The queue of async waitetrs
	private LinkedList<AsyncWaiter> asyncWaiters;

	//  Completed futures used to return true and false results
	private static final CompletableFuture<Boolean> trueFuture = CompletableFuture.completedFuture(true);
	private static final CompletableFuture<Boolean> falseFuture = CompletableFuture.completedFuture(false);;
	    
	/**
     * Constructor
     */
    public CountDownLatchAsync(int initialCount) {
		if (initialCount < 0)
			throw new IllegalArgumentException("initialCount");
		this.initialCount = initialCount;
		count = new AtomicInteger(initialCount);
		// If the count down latch is initialized as closed, initialize the wait list.
		if (count.get() > 0)
        	asyncWaiters = new LinkedList<AsyncWaiter>();
	}
	
    /**
	 * Asynchronous Task-based Asynchronous Pattern (TAP) interface.
	 */

    /**
	 * Wait asynchronously for the latch to open enabling, optionally, a timeout
	 * and/or cancellation.
	 */
    private CompletableFuture<Boolean> doAwaitAsync(boolean timed, long timeout, TimeUnit unit) {
		// We get the count value with a "volatile read", so the visibility is guaranteed
		if (count.get() == 0)
			return trueFuture;
		synchronized(theLock) {
			// After acquire the lock we must re-check the latch state, because
			// however it may have been opened by another thread.
			if (count.get() == 0)
				return trueFuture;

            // If the wait was specified as immediate, return failure
            if (timed && timeout == 0)
				return falseFuture;
			
			// Create a request node and insert it in the async waiters queue
			AsyncWaiter awaiter = new AsyncWaiter();
			asyncWaiters.addLast(awaiter);
		
			/**
			 * If a timeout was specified, start a timer.
			 * Since that all paths of code that cancel the timer execute on other
			 * threads and must aquire the lock, we has the guarantee that the field
			 * "acquirer.timer" is correctly set when the method AsyncAcquire.close()
			 * is called.
			 */
			if (timed)
				awaiter.timer = Delayer.delay(awaiter, timeout, unit);
			return awaiter;
		}
	}

	/**
	 * Wait until latch opens asynchronously unconditionally.
	 */
	public CompletableFuture<Boolean> awaitAsync() {
		return doAwaitAsync(false, 0L, null);
	}

	/**
	 * Wait until latch opens asynchronously enabling the timeout.
	 */
	public CompletableFuture<Boolean> awaitAsync(long timeout, TimeUnit unit) {
		return doAwaitAsync(true, timeout, unit);
	}

	/**
	 * Returns the latch state
	 */
	public boolean isOpen() { return open; }

    /**
	 *	Synchronous interface implemented using the asynchronous TAP interface.
	 */
	
	 /**
	 * Try to cancel an asynchronous request identified by its CF<>.
	 */
	boolean tryCancelAwaitAsync(CompletableFuture<Boolean> awaiterFuture) {
		AsyncWaiter awaiter = (awaiterFuture instanceof AsyncWaiter) ? (AsyncWaiter)awaiterFuture : null;
		if (awaiter == null)
			throw new IllegalArgumentException("awaiterFuture");
		if (awaiter.tryLock()) {
			synchronized(theLock) {
				asyncWaiters.remove(awaiter);	// no operation if object is not in the list
			}
			// Release resources and complete the CF<>
			awaiter.close();
			awaiter.completeExceptionally(new CancellationException());
			return true;
		}
		return false;
	}

    /**
	 * Wait synchronously for the latch to open enabling, optionally,
	 * timeout and/or cancellation.
	 */
    private boolean doAwait(boolean timed, long timeout, TimeUnit unit) throws InterruptedException {
		CompletableFuture<Boolean> awaitFuture = doAwaitAsync(timed, timeout, unit); 
		try {
            return awaitFuture.get();
        } catch (InterruptedException ie) {
			// Try to cancel the async await
			if (tryCancelAwaitAsync(awaitFuture))
				throw ie;
			
			// Here, we known that the request was already completed.
			// Return the underlying result, filtering any possible interrupts.
			try {
				do {
					try {
						return awaitFuture.get();
					} catch (InterruptedException ie2) {
						// While waiting for result, we filter all interrupts
					} catch (Throwable ex2) {
						// We never get here, because we never complete the CF<> exceptionally.
					}
				} while (true);
            } finally {
				// Anyway, re-assert the interrupt
                Thread.currentThread().interrupt();
            }
        } catch (Throwable ex) {
			// We never get here, because we never complete the CF<> exceptionally.
		}
		return false;
	}

	/**
	 * Wait until latch opens synchronously unconditionally.
	 */
	public boolean await() throws InterruptedException {
		return doAwait(false, 0L, null);
	}

	/**
	 * Wait until latch opens synchronously enabling the timeout.
	 */
	public boolean await(long timeout, TimeUnit unit) throws InterruptedException {
		return doAwait(true, timeout, unit);
	}
	
	/**
	 * Registers one or more signals with the CountdownLatch, decrementing the
	 * value of CurrentCount by the specified amount.
	 */
	public boolean signal(int signalCount) {
		if (signalCount < 1)
			throw new IllegalArgumentException("signalCount");
		int c;
		do {
			c = count.get();
			if (c == 0 || signalCount > c)
				throw new  IllegalStateException();
		} while (!count.compareAndSet(c, c - signalCount));
		if (c > signalCount)
			return false;
		
		// The latch is now open, release all async waiters.
		// A list to hold temporarily the async waiters to complete later
		// after release the lock.

		// A list to hold temporarily the async waiters to complete
		// later without owning the lock.
		LinkedList<AsyncWaiter> completed = null;
		synchronized(theLock) {
			if (asyncWaiters.size() > 0)
				completed = asyncWaiters;
			asyncWaiters = null;
		}
		// Complete the tasks of the async waiters without owning the lock
		if (completed != null) {
			for (AsyncWaiter awaiter : completed) {
				awaiter.close();
				awaiter.complete(true);
			}
		}
		return true;
	}

	public boolean signal() { return signal(1); }

	/**
	 * Increments the CurrentCount by a specified value.
	 */
	public void addCount(int signalCount) {
		if (signalCount < 1)
			throw new IllegalArgumentException("signalCount");
		int c;
		do {
			c = count.get();
			if (c == 0 || c + signalCount < c)
				throw new IllegalStateException();
		} while (!count.compareAndSet(c, c + signalCount));
	}

	public void  addCount() { addCount(1); }

	/**
	 * Attempts to increment CurrentCount by a specified value.
	 */
	public boolean tryAddCount(int signalCount) {
		if (signalCount < 1)
			throw new IllegalArgumentException("signalCount");
		int c;
		do {
			c = count.get();
			if (c == 0)
				return false;
			if (c + signalCount < c)
				throw new IllegalStateException();
		} while (!count.compareAndSet(c, c + signalCount));
		return true;
	}

	public boolean tryAddCount() { return tryAddCount(1); }

	/**
	 * Gets the number of remaining signals required to open the latch.
	 */ 
	public int getCurrentCount() { return count.get(); }

	/**
	 * Gets the numbers of signals initially required to set the latch.
	 */
	public int getInitialCount() { return initialCount; }

	/**
	 * Indicates whether the count down latch's current count has reached zero.
	 */
	public boolean isSet() { return count.get() == 0; }

	/**
	 * Entry point to run tests
	 */

	public static void main(String[] args) throws InterruptedException {
		CountDownLatchAsyncTests.testWaitAsync();	}

}

/**
 * Test code
 */
class CountDownLatchAsyncTests {
	static final int SETUP_TIME = 50;
	static final int UNTIL_OPEN_TIME = 500;
	static final int THREAD_COUNT = 10;
	static final int EXIT_TIME = 100;
	static final int WAIT_ASYNC_TIMEOUT = 100;

	static void Log(String msg) {
		System.out.printf("[#%02d]: %s\n", Thread.currentThread().getId(), msg);
	}

	public static void testWaitAsync() throws InterruptedException {
		CountDownLatchAsync cdl = new CountDownLatchAsync(1);
		Thread[] waiters = new Thread[THREAD_COUNT];
		boolean timed = false;
		long timeout = WAIT_ASYNC_TIMEOUT /* + UNTIL_OPEN_TIME */;

		for (int i = 0; i < THREAD_COUNT; i++) {
			int li = i;

			waiters[i] = new Thread(() -> {
				Log(String.format("--[#%02d]: waiter thread started", li));
				if (li > 0)
					cdl.addCount();
				try {
					CompletableFuture<Boolean> awaitFuture;
					if (timed)
						awaitFuture = cdl.awaitAsync(timeout, TimeUnit.MILLISECONDS);
					else
						awaitFuture = cdl.awaitAsync();
					Log(String.format("--[#%02d]: returned from async await", li));
					try {
						if (awaitFuture.get())
							Log(String.format("--[#%02d]: count down latch opened", li));
						else
							Log(String.format("--[#%02d]: awaitAsync() timed out", li));
					} catch (ExecutionException ee) {
						// We never get here
					}
				} catch (InterruptedException ie) {
					Log(String.format("--[#%02d]: awaiter thread was interrupted"));
				}		
			});
			waiters[i].start();
		}

		Thread.sleep(SETUP_TIME + UNTIL_OPEN_TIME);

		for (int i = 0; i < THREAD_COUNT; i++)
			cdl.signal();
		
		Thread.sleep(EXIT_TIME);

		for (int i = 0; i < THREAD_COUNT; i++) {
			if (waiters[i].isAlive())
				waiters[i].interrupt();
			waiters[i].join();
		}
		Log("--test terminated");
	}
}

File: Exams/Resoluções de Exames by Carlos Martins/synchs-async/CountDownLatchAsync.cs
/**
 *
 *  ISEL, LEIC, Concurrent Programming
 *
 *  Count down latch with asynchronous and synchronous interfaces
 *
 *  Carlos Martins, December 2019
 *
 **/

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Diagnostics;

public class CountDownLatchAsync {

		// Type used to represent each asynchronous waiter
	private class AsyncWaiter: TaskCompletionSource<bool> {
		internal readonly CancellationToken cToken;
		internal CancellationTokenRegistration cTokenRegistration;
		internal Timer timer;
		const int PENDING = 0, LOCKED = 1;
		private volatile int _lock; 	// The waiter lock
		
		internal AsyncWaiter(CancellationToken cToken) : base() {
			this.cToken = cToken;
			this._lock = PENDING;
		}

		/**
		 * Tries to lock the waiter in order to satisfy or cancel it.
		 */
		internal bool TryLock() {
			return _lock == PENDING &&
				   Interlocked.CompareExchange(ref _lock, LOCKED, PENDING) == PENDING;
		}

		/**
		 * Disposes resources associated with this async waiter.
		 *
		 * Note: when this method is called we are sure that the fields "timera"
		 *       and "cTokenRegistration" are correctly affected
		 */
		internal void Dispose(bool canceling) {
			// The CancellationTokenRegistration is disposed off after the cancellation
			// handler is called.
			if (!canceling && cToken.CanBeCanceled)
				cTokenRegistration.Dispose();
			timer?.Dispose();
		}
	}

	// The lock - we do not use the monitor functionality
	private readonly object theLock = new object();

	// The initial and current count
	private readonly int initialCount;
	private volatile int count;

	// The queue of async waitetrs
	private LinkedList<AsyncWaiter> asyncWaiters;

	// Pre-initialized delegates that are used as cancellation handlers
    private readonly Action<object> cancellationHandler;
	private readonly TimerCallback timeoutHandler;

    // Pre-initialized completed task to return "true" and "false" results.
    private static readonly Task<bool> trueTask = Task.FromResult<bool>(true);
	private static readonly Task<bool> falseTask = Task.FromResult<bool>(false);

	/**
	 * Constructor
	 */
	public CountDownLatchAsync(int initialCount = 1) {
		if (initialCount < 0)
			throw new ArgumentOutOfRangeException("initialCount");
		this.count = this.initialCount = initialCount;
		if (count > 0) {
	    	// If the latch is initialized as closed, initialize its fields
        	cancellationHandler = new Action<object>((awaiterNode) => WaitCancellationHandler(awaiterNode, true));
        	timeoutHandler = new TimerCallback((awaiterNode) => WaitCancellationHandler(awaiterNode, false));
        	asyncWaiters = new LinkedList<AsyncWaiter>();
		}
    }

	/**
	 * Auxiliary methods
	 */

	/**
	 * Common code when cancelling an async waiter.
	 */
    private void WaitCancellationHandler(object _awaiterNode, bool canceling) {
		LinkedListNode<AsyncWaiter> awaiterNode = (LinkedListNode<AsyncWaiter>)_awaiterNode;
		AsyncWaiter awaiter = awaiterNode.Value;
		if (awaiter.TryLock()) {
			// Acquire the lock and remove the async waiter from the waiters list
			lock(theLock) {
				if (awaiterNode.List != null)
					asyncWaiters.Remove(awaiterNode);
			}
			awaiter.Dispose(canceling);
			if (canceling)
				awaiter.SetCanceled();
			else
				awaiter.SetResult(false);
		}
	}

    /**
	 * Asynchronous Task-based Asynchronous Pattern (TAP) interface.
	 */

    /**
	 * Wait asynchronously for the count down latch to open enabling, optionally,
	 * a timeout and/or cancellation.
	 */
    public Task<bool> WaitAsync(int timeout = Timeout.Infinite,
							    CancellationToken cToken = default(CancellationToken)) {
		// The "count" field is volatile, so the visibility is guaranteed
		if (count == 0)
			return trueTask;
		lock(theLock) {
			// After acquire the lock we must re-check the count down latch state,
			// because however it may have been opened by another thread.
			if (count == 0)
				return trueTask;

            // If the wait was specified as immediate, return failure
            if (timeout == 0)
				return falseTask;
			
			// If a cancellation was already requested return a task in the Canceled state
			if (cToken.IsCancellationRequested)
				return Task.FromCanceled<bool>(cToken);
						
			// Create a request node and insert it in requests queue
			AsyncWaiter awaiter = new AsyncWaiter(cToken);
			LinkedListNode<AsyncWaiter> awaiterNode = asyncWaiters.AddLast(awaiter);
		
			/**
			 * Activate the specified cancelers owning the lock.
			 */
			
			/**
			 * Since the timeout handler, that runs on a thread pool's worker thread,
			 * acquires the lock before access the "request.timer" and "request.cTokenRegistration"
			 * these assignements will be visible to timer handler.
			 */
			if (timeout != Timeout.Infinite)
				awaiter.timer = new Timer(timeoutHandler, awaiterNode, timeout, Timeout.Infinite);
			
			/**
			 * If the cancellation token is already in the canceled state, the cancellation
			 * delegate will run immediately and synchronously, which *causes no damage* because
			 * this processing is terminal and the implicit locks can be acquired recursively.
			 */
			if (cToken.CanBeCanceled)
            	awaiter.cTokenRegistration = cToken.Register(cancellationHandler, awaiterNode);
	
			// Return the Task<bool> that represents the asynchronous operation
			return awaiter.Task;
		}
    }

	/**
	 *	Synchronous interface implemented using the asynchronous TAP interface.
	 */

	/**
	 * Try to cancel an asynchronous request identified by its task.
	 */
	private bool CancelWaitByTask(Task<bool> awaiterTask) {
        AsyncWaiter awaiter = null;
        lock(theLock) {
			foreach (AsyncWaiter _awaiter in asyncWaiters) {
				if (_awaiter.Task == awaiterTask) {
                    if (_awaiter.TryLock()) {
						awaiter = _awaiter;
                        asyncWaiters.Remove(_awaiter);
					}
					break;
				}
			}
		}
		if (awaiter != null) {
			// Dispose resources and complete async wait with TaskCanceledException
            awaiter.Dispose(false);
            awaiter.SetCanceled();
        }
		return awaiter != null;
	}


    /**
	 * Wait synchronously for the count down latch to open enabling, optionally,
	 * timeout and/or cancellation.
	 */
    public bool Wait(int timeout = Timeout.Infinite,
					 CancellationToken cToken = default(CancellationToken)) {
		Task<bool> waitTask = WaitAsync(timeout, cToken); 
		try {
            return waitTask.Result;
        } catch (ThreadInterruptedException) {
			// Try to cancel the asynchronous request
			if (CancelWaitByTask(waitTask))
				throw;
			// We known that the request was already completed or cancelled,
			// return the underlying result
			try {
				do {
					try {
						return waitTask.Result;
					} catch (ThreadInterruptedException) {
						// ignore interrupts while we wait for the task's result
					} catch (AggregateException ae) {
                		throw ae.InnerException;
					}
				} while (true);
            } finally {
				// Anyway, re-assert the interrupt
                Thread.CurrentThread.Interrupt();
            }
        } catch (AggregateException ae) {
            throw ae.InnerException;
        }
	}

	/**
	 * Complete the tasks associated to the completed async waiters.
	 *  Note: This method is called when calling thread does not own the lock.
	 */
    private void CompleteAsyncWaiters(LinkedList<AsyncWaiter> completed) {
        if (completed != null) {
            foreach (AsyncWaiter awaiter in completed) {
				if (awaiter.TryLock()) {
					awaiter.Dispose(false);
                	awaiter.SetResult(true);
				}
            }
        }
    }

	/**
	 * Registers one or more signals with the CountdownLatch, decrementing the
	 * value of CurrentCount by the specified amount.
	 */
	public bool Signal(int signalCount = 1) {
		if (signalCount < 1)
			throw new ArgumentOutOfRangeException("signalCount");
		int c;
		do {
			c = count;
			if (c == 0 || signalCount > c)
				throw new InvalidOperationException();
		} while (Interlocked.CompareExchange(ref count, c - signalCount, c) != c);
		if (c > signalCount)
			return false;
		
		// The latch is now open, release all async waiters.
		// A list to hold temporarily the async waiters to complete later
		// after release the lock.
	    LinkedList<AsyncWaiter> toComplete = null;
		lock(theLock) {
			if (asyncWaiters.Count > 0)
				toComplete = asyncWaiters;
			asyncWaiters = null;
		}
		CompleteAsyncWaiters(toComplete);
		return true;
	}

	/**
	 * Increments the CurrentCount by a specified value.
	 */
	public void AddCount(int signalCount = 1) {
		if (signalCount < 1)
			throw new ArgumentOutOfRangeException("signalCount");
		int c;
		if (signalCount <= 0)
			throw new ArgumentOutOfRangeException("signalCount");
		do {
			c = count;
			if (c == 0 || c + signalCount < c)
				throw new InvalidOperationException();
		} while (Interlocked.CompareExchange(ref count, c + signalCount, c) != c);
	}

	/**
	 * Attempts to increment CurrentCount by a specified value.
	 */
	public bool TryAddCount(int signalCount = 1) {
		int c;
		do {
			c = count;
			if (c == 0)
				return false;
			if (c + signalCount < c)
				throw new InvalidOperationException();
		} while (Interlocked.CompareExchange(ref count, c + signalCount, c) != c);
		return true;
	}

	/**
	 * Gets the number of remaining signals required to open the latch.
	 */ 
	public int CurrentCount { get { return count; } }

	/**
	 * Gets the numbers of signals initially required to set the latch.
	 */
	public int InitialCount { get { return initialCount; } }

	/**
	 * Indicates whether the count down latch's current count has reached zero.
	 */
	public bool IsSet {
		get { return count == 0; }
	}
}

/**
 * Test code
 */

internal class CountDownLatchAsyncTests {
	const int SETUP_TIME = 50;
	const int UNTIL_OPEN_TIME = 500;
	const int THREAD_COUNT = 10;
	const int EXIT_TIME = 100;
	const int WAIT_ASYNC_TIMEOUT = 2000;


	internal static void Log(String msg) {
		Console.WriteLine($"[#{Thread.CurrentThread.ManagedThreadId:D2}]: {msg}");
	}

	private static void TestWaitAsync() {
		CountDownLatchAsync cdl = new CountDownLatchAsync(1);
		CancellationTokenSource cts = new CancellationTokenSource();
		Thread[] waiters = new Thread[THREAD_COUNT];
		//int timeout = Timeout.Infinite;
		int timeout = WAIT_ASYNC_TIMEOUT;

		for (int i = 0; i < THREAD_COUNT; i++) {
			int li = i;

			waiters[i] = new Thread(() => {
				Log($"--[#{li}:D2]: waiter thread started");
				if (li > 0)
					cdl.TryAddCount(1);
				try {
					var waitTask = cdl.WaitAsync(timeout: timeout, cToken: cts.Token);
					Log($"--[#{li:D2}]: returned from async wait");
					if (waitTask.Result)
						Log($"--[#{li:D2}]: count down latch opened");
					else
						Log($"--[#{li:D2}]: WaitAsync() timed out");
				} catch (AggregateException ae) {
					Log($"***[#{li:D2}]: {ae.InnerException.GetType().Name}: {ae.InnerException.Message}");
				} catch (ThreadInterruptedException) {
					Log($"***[#{li:D2}]: thread was interrupted");
				}
				Log($"--[#{li:D2}]: waiter thread exiting");
			});
			waiters[i].Start();
		}

		Thread.Sleep(SETUP_TIME + UNTIL_OPEN_TIME);
		Log("--press <enter> to signal");
		Console.ReadLine();
		//cts.Cancel();
		for (int i = 0; i < THREAD_COUNT; i++)
			cdl.Signal();

		Thread.Sleep(EXIT_TIME);

		for (int i = 0; i < THREAD_COUNT; i++) {
			if (waiters[i].IsAlive)
				waiters[i].Interrupt();
			waiters[i].Join();
		}
		Log("--test terminated");
	}

    static void Main() {
		TestWaitAsync();
    }
}


